### ISSUE TITLE
Disable auth if `{DAV:}unauthenticated`

### ISSUE THREAD

**Issue by Hywan** [fruux/sabre-dav#657]
When ACL contains `{DAV:}unauthenticated` or `{DAV:}all` we should allow the ability to perform these actions without requiring authentication.

The way this needs to work is as follows:

1. The authentication plugin now automatically calls `challenge` on every authentication backend. The plugin should get some kind of flag that allows someone to disable this. Lets call it `noAutoChallenge` for the moment.
2. The acl plugin should automatically look for the authentication plugin, and set the `noAutoChallenge` to true.
3. When the ACL plugin is required to check for privileges (`checkPrivileges` method), and the check _fails_, check if *any* user is currently logged in. If no user is logged in, call the `challenge` method manually on the authentication plugin.

In short: we postpone the challenge *until* the first time an ACL check fails.

Risks
-------

The risk is that *before* this change, *any* action would require authentication, and after this change this is no longer true. There's a possibility that there's things within sabre/dav that don't correctly check the access control system before doing something.

If this is true, we now open up these requests, allowing people to perform these actions without requiring proper authentication. This could be bad.

Even if sabre/dav code doesn't do this, other people that use sabre/dav might.

**Comment by mcassaniti** [comment]
I know that a use case to go with this would be highly ideal, so here it is. What happens when you want your calendaring system to provide free/busy information via a public URL **without** authentication. At present it is expected that authentication is uesd throughout the CalDAV backends exclusively, and I would like to see at least one backend created there that allows for a free/busy public URL (something like freebusy.php/user@domain).

I know there's currently no plugin available to do a public free/busy, but I wouldn't even want to start without optional authentication.

**Comment by evert** [comment]
Hi! That's a nice use-case indeed. Would be a good separate feature as well.

**Comment by staabm** [comment]
> The risk is that before this change, any action would require authentication, and after this change this is no longer true. There's a possibility that there's things within sabre/dav that don't correctly check the access control system before doing something.

Instead of allowing everything and add a blacklist for all "secured" resources you could go the other way around and let the user configure whitelisted actions which dont require auth. So the risk to make things public which shouldn't be won't be a problem anymore.

**Comment by evert** [comment]
@staabm to a certain extent that _is_ how ACL works. Everything is disabled by default, unless somebody turned it on. A problem is that somebody might have added support for a new action, but not integrated with acl.

There's no easy way to prevent this :S
